### PR TITLE
Skill to add unit tests to a dbt project

### PR DESCRIPTION
**Pull request by dbeatty10**
resolves #

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-oss-template/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-oss-template/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,902 @@
+---
+name: adding-dbt-unit-test
+description: Use when adding unit tests for a dbt model
+---
+
+# Add unit test for a dbt model
+
+## Overview
+
+Historically, dbt's test coverage was confined to data tests, assessing the quality of input data or resulting datasets' structure. However, these tests could only be executed _after_ building a model.
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]: I'm a little skeptical of _starting_ with the history lesson, although I do think this will be valuabe since models probably don't know that much about dbt unit tests. Let's start with a general "what are unit tests in dbt"

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,902 @@
...
+Historically, dbt's test coverage was confined to data tests, assessing the quality of input data or resulting datasets' structure. However, these tests could only be executed _after_ building a model. 
+
+There is an additional type of test to dbt - unit tests. In software programming, unit tests validate small portions of your functional code, and they work much the same way here. Unit tests allow you to validate your SQL modeling logic on a small set of static inputs _before_ you materialize your full model in production. Unit tests enable test-driven development, benefiting developer efficiency and code reliability. 
+
+Unit tests allow enforcing that all the unit tests for a model pass before it is materialized (i.e. dbt won't materialize the model in the database if *any* of its unit tests do not pass).
+
+## When to use
+
+You should unit test a model:
+- Adding Model-Input-Output scenarios for the general case and edge cases to prevent regressions if the model logic is changed at a later date.
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]:
```SUGGESTION
- Adding Model-Input-Output scenarios for the intended functionality of the model as well as edge cases to prevent regressions if the model logic is changed at a later date.

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,902 @@
...
+Cases we don't recommend creating unit tests for:
+- Built-in functions that are tested extensively by the warehouse provider. If an unexpected issue arises, it's more likely a result of issues in the underlying data rather than the function itself. Therefore, fixture data in the unit test won't provide valuable information.
+    - common SQL spec functions like `min()`, etc.
+
+## General format
+
+dbt unit test uses a trio of the model, given inputs, and expected outputs (Model-Inputs-Outputs):
+
+1. `model` - when building this model
+2. `given` inputs - given a set of source, seeds, and models as preconditions
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:45]:
```suggestion
2. `given` inputs - given a set of source, seeds, macros and models as preconditions
```

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,902 @@
...
+
+### Workflow
+
+### 1. Choose the model to test
+
+Self explanatory -- the title says it all!
+
+### 2. Mock the inputs
+
+Create an input for each of the nodes the model depends on. Specify the mock data it should use. The mock data only needs include the subset of columns used within this test case. Specify the `format` if different than the default (YAML `dict`). See the "Data `format`s for unit tests" section below to determine which `format` to use.
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]: Can we make these (and similar sections) bulleted

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,902 @@
...
+    # Output
+    expect:
+      rows:
+        - {
+            order_id: 76,
+            count_drink_items: 0,
+          }
+```
+
+## Another example of unit testing a model
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]: At this point more examples should probably be moved to a different file in the skill.

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,902 @@
...
+## Handy to know
+
+- Unit tests must be defined in a YAML file in your `model-paths` directory (`models/` by default)
+- Fixture files for unit tests must be defined in a SQL or CSV file in your `test-paths` directory (`tests/fixtures` by default)
+- Include all `ref` or `source` model references in the unit test configuration as `input`s to avoid "node not found" errors during compilation.
+- If your model has multiple versions, by default the unit test will run on *all* versions of your model.
+- If you want to unit test a model that depends on an ephemeral model, you must use `format: sql` for the ephemeral model input.
+- Table names within the model must be aliased in order to unit test `join` logic
+
+## YAML for specifying unit tets
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:196]: This feels like a different file in the skill

Comment (jasnonaz) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md:196]: reference/spec or something like that

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,902 @@
...
+...
+
+```
+</File>
+
+# Data `format`s for unit tests
+
+dbt supports three formats for mock data within unit tests:
+
+1. `dict` (default): Inline YAML dictionary values.
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:240]: Since dict is default, can we add a short description of why they might want to do csv or sql and then move the examples to another file

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,902 @@
...
+1,gerda
+2,michelle
+
+```
+</File>
+
+## `sql`
+
+Using this format:
+- Provides more flexibility for the unit testing column that have a data type not supported by the `dict` or `csv` formats
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]: THIS is great - this is the stuff we keep

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,902 @@
...
+select 2 as id, 'michelle', null as loaded_at as name
+
+```
+</File>
+
+**Note:** Contrary to dbt SQL models, Jinja is unsupported within SQL fixtures for unit tests.
+
+# Special cases
+
+## Unit testing incremental models
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]: incremental and versioned examples should live in their own file with callouts that these are special cases and the model should look into it. For the most part - keep the prose,, move the code.

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,902 @@
...
+      overrides:
+        macros:
+          # explicity set star to relevant list of columns
+          dbt_utils.star: col_a,col_b,col_c 
+      ...
+
+  ``` 
+</File>
+
+#### Platform/adapter-specific caveats
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:272]: These should mostly be moved to platform specific examples and content here should read something like:

"There are platform specific details required if implementing on (redshift, etc). If you are on that adapter, read the platform_name_caveats.md file.

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,902 @@
...
+→  ,cool@example.com,True→False
+   ,cool@unknown.com,False
+```
+
+There are two main possibilities when a unit test fails:
+
+1. There was an error in the way the unit test was constructed (false positive)
+2. There is an bug is the model (true positive)
+
+It takes expert judgement to determine one from the other.
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:313]: Can we have some guidance on how to determine one from the other?

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,902 @@
...
+
+Use the `--empty` flag to build an empty version of the models to save warehouse spend. 
+
+```bash
+
+dbt run --select "stg_customers top_level_email_domains" --empty
+
+```
+
+### Fixture files
```
Comment (jasnonaz) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]: Is this needed given the above section talking about files?

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 1 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md]: Resolved in 1abb0744fc9dce883f7c157502c10ad2f707d4cc

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 3 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md:45]: Since macros go under `overrides` rather than under `given`, I'm not going to commit this as-is.

But good point that macros fall under the inputs / pre-conditions umbrella (along with project variables and environment variables).

So we'll either want to add that info here or just leave it to the later sections that cover those scenarios

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 4 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md]: Addressed in 4d0bc0fba77996307c423d0855bc9efa9748b739

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 6 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md:196]: Addressed in d27a66d403d1e69549fc2941d8fa032a07b79426

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 5 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md]: Addressed in a4c2dbb3c97cecd1eaeb7be57e8cd474d07037f7

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 7 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md:240]: Addressed in 62383ed20854106a632148fb8b6074f878406ad8 and 1b5224a47ba738e49f41f294a2550cb928f25ddb

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 10 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md:272]: Addressed in ea5c89a4eb85c4e96035098b53710e341ed47274

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 12 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md]: Partially addressed in 5719ff46f1fc0a2eb8e4a4a8b061c985585768a3.

Can revisit to consolidate any duplicated content.

File: dbt-unit-tests/skills/warehouses/bigquery/caveats.md [hunk 13]
```diff
@@ -0,0 +1,3 @@
+# Caveats for BigQuery
```
Comment (b-per) [inline dbt-unit-tests/skills/warehouses/bigquery/caveats.md:1]: I really think we need this type of info! Thanks for adding.

My comments:
- is caveat the correct word here? if we add recommendations and best practices for other parts in dbt for BQ, those might not be caveats. I like `instructions_bigquery.md` or something like that
- we could add all of those under a specific folder for adapter/warehouse specific instructions
- we should call out the name of the file in any skill that would need to get this information, otherwise the LLM will not look at it

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,532 @@
+---
+name: adding-dbt-unit-test
+description: Use when adding unit tests for a dbt model
+---
+
+# Add unit test for a dbt model
+
+## What are unit tests in dbt
+
+In software programming, unit tests validate small portions of your functional code, and they work much the same way in dbt. dbt uwnit tests allow you to validate your SQL modeling logic on a small set of static inputs _before_ you materialize your full model in production. dbt unit tests enable test-driven development, benefiting developer efficiency and code reliability.
```
Comment (b-per) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md]:
```suggestion
In software programming, unit tests validate small portions of your functional code, and they work much the same way in dbt. dbt unit tests allow you to validate your SQL modeling logic on a small set of static inputs _before_ you materialize your full model in production. dbt unit tests enable test-driven development, benefiting developer efficiency and code reliability.
```

File: dbt-unit-tests/skills/add-unit-test/spec.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
+See below for all the required and optional keys in the YAML definition of unit tests.
```
Comment (b-per) [inline dbt-unit-tests/skills/add-unit-test/spec.md:1]: Yes to this type of info!!!

This is the information dense type of content that is maybe too dense for humans to reason through, but it is exactly what LLMs need!

File: dbt-unit-tests/skills/add-unit-test/examples.md [hunk 16]
```diff
@@ -0,0 +1,214 @@
+## Another example of unit testing a model
+
+This example creates a new `dim_customers` model with a field `is_valid_email_address` that calculates whether or not the customer’s email is valid: 
+
+<file name='dim_customers.sql'>
```
Comment (b-per) [inline dbt-unit-tests/skills/add-unit-test/examples.md]: We haven't used this `<file ...> syntax in the other skills. Open question, should we do it everywhere or should we not use it? I think we should be consistent though.

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 9 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md]: Addressed in e395b1d1474b8b1690e2182e357814e7f95f66f6

File: dbt-unit-tests/skills/warehouses/bigquery/caveats.md (hunk 13 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/warehouses/bigquery/caveats.md:1]: > we should call out the name of the file in any skill that would need to get this information, otherwise the LLM will not look at it

✅ Done!

> is caveat the correct word here? if we add recommendations and best practices for other parts in dbt for BQ, those might not be caveats. I like instructions_bigquery.md or something like that

One definition of caveat is "specific stipulations, conditions, or limitations", so it is a good descriptor in this case. We can call it whatever we want though, and I don't feel strongly either way.

> we could add all of those under a specific folder for adapter/warehouse specific instructions

Are you thinking something like this within a subfolder with simply the name of the adapter/warehouse?

```
my-skill/
├── SKILL.md
└── adapter/
    ├── bigquery.md (adapter/warehouse specific instructions - loaded when needed)
    ├── databricks.md (adapter/warehouse specific instructions - loaded when needed)
    ├── redshift.md (adapter/warehouse specific instructions - loaded when needed)
    └── snowflake.md (adapter/warehouse specific instructions - loaded when needed)
```

Or flat like this?

```
my-skill/
├── SKILL.md
├── bigquery.md (adapter/warehouse specific instructions - loaded when needed)
├── databricks.md (adapter/warehouse specific instructions - loaded when needed)
├── redshift.md (adapter/warehouse specific instructions - loaded when needed)
└── snowflake.md (adapter/warehouse specific instructions - loaded when needed)
```

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 17]
```diff
@@ -0,0 +1,532 @@
...
+Unit tests allow enforcing that all the unit tests for a model pass before it is materialized (i.e. dbt won't materialize the model in the database if *any* of its unit tests do not pass).
+
+## When to use
+
+You should unit test a model:
+- Adding Model-Input-Output scenarios for the intended functionality of the model as well as edge cases to prevent regressions if the model logic is changed at a later date.
+- Verifying that a bug fix solves a bug report for an existing dbt model.
+
+More examples:
+- When your SQL contains complex logic:
```
Comment (b-per) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:21]: Should we add complex joins logic?

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 18]
```diff
@@ -0,0 +1,532 @@
...
+2. `given` inputs - given a set of source, seeds, and models as preconditions
+3. `expect` output - then expect this row content of the model as a postcondition
+
+### Workflow
+
+### 1. Choose the model to test
+
+Self explanatory -- the title says it all!
+
+### 2. Mock the inputs
```
Comment (b-per) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:54]: Should we recommend using `dbt show` to explore the existing inputs/outputs and sanitize them to make sure they don't contain sensitive data?

File: dbt-unit-tests/skills/add-unit-test/SKILL.md [hunk 19]
```diff
@@ -0,0 +1,532 @@
+---
+name: adding-dbt-unit-test
+description: Use when adding unit tests for a dbt model
```
Comment (b-per) [inline dbt-unit-tests/skills/add-unit-test/SKILL.md:3]: We'd want the SKILL file to be shorter I think. The max recommended is 500 lines. We can split into other files.

File: dbt-unit-tests/skills/warehouses/bigquery/caveats.md (hunk 13 above)
Comment (b-per) [reply dbt-unit-tests/skills/warehouses/bigquery/caveats.md:1]: I was actually thinking of a third option (close to the 1st one)

```
my-skill1/
└── SKILL.md
my-skill2/
└── SKILL.md
adapter/
    ├── bigquery.md (adapter/warehouse specific instructions - loaded when needed)
    ├── databricks.md (adapter/warehouse specific instructions - loaded when needed)
    ├── redshift.md (adapter/warehouse specific instructions - loaded when needed)
    └── snowflake.md (adapter/warehouse specific instructions - loaded when needed)
```

The adapter specific things we'd have to say will be short enough that we can have 1 file to cover all the skills.

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 19 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md:3]: Got it down to ~300 lines after splitting content into other files

File: dbt-unit-tests/skills/add-unit-test/SKILL.md (hunk 18 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/SKILL.md:54]: This is a fantastic idea!

Might even be a good stand-alone and reusable skill.

File: dbt-unit-tests/skills/warehouses/bigquery/caveats.md (hunk 13 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/warehouses/bigquery/caveats.md:1]: Addressed in 2542c0f8dfc150e41fb0d76782806fd0ba55827e, fdaf6abfa2e40a60b4625c7d0a7cd63b84770b89, and dd24b72d0e6921ed9a1fe2eb471f40b1be4b5ffb.

We can move it around / rename it further as-needed.

File: dbt-unit-tests/skills/add-unit-test/examples.md (hunk 16 above)
Comment (dbeatty10) [reply dbt-unit-tests/skills/add-unit-test/examples.md]: Updated to use the following syntax within bcf42928cc962a26371cf191ec71586f6c7866de:

\`models/schema.yml\`

\```yaml
    # stuff
\```
